File: 2020/4/solution.py
import re
import os

REQUIRED_FIELDS = {'byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid'}
VALID_ECL = ('amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth')

class Passport:
  def __init__(self) -> None:
    self.fields = {}

  def add_field(self, field: str) -> None:
    k, v = field.split(':')
    self.fields[k] = v

  def validate(self, part2 = False) -> bool:
    if not REQUIRED_FIELDS.issubset(self.fields.keys()):
      return False
    if part2:
      if not re.match(r'^\d{4}$', self.fields['byr']):
        return False
      if not 1920 <= int(self.fields['byr']) <= 2002:
        return False
      if not re.match(r'^\d{4}$', self.fields['iyr']):
        return False
      if not 2010 <= int(self.fields['iyr']) <= 2020:
        return False
      if not re.match(r'^\d{4}$', self.fields['eyr']):
        return False
      if not 2020 <= int(self.fields['eyr']) <= 2030:
        return False
      hgt_match = re.match(r'^(\d+)(cm|in)$', self.fields['hgt'])
      if not hgt_match:
        return False
      hgt, unit = int(hgt_match.group(1)), hgt_match.group(2)
      if unit == 'cm' and not 150 <= hgt <= 193:
        return False
      if unit == 'in' and not 59 <= hgt <= 76:
        return False
      if not re.match(r'^#[0-9a-f]{6}$', self.fields['hcl']):
        return False
      if self.fields['ecl'] not in VALID_ECL:
        return False
      if not re.match(r'^\d{9}$', self.fields['pid']):
        return False
    return True

if __name__ == "__main__":
  with open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'input.txt'),'r') as file:
    lines = [line.strip() for line in file]

  pattern = r'(\w{3}:\S+)'

  passports: list[Passport] = []
  passport = Passport()
  for line in lines:
    if not line:
      passports.append(passport)
      passport = Passport()
    else:
      matches = re.findall(pattern, line)
      for match in matches:
        passport.add_field(match)

  part1 = sum([passport.validate() for passport in passports])
  print(part1)
  part2 = sum([passport.validate(True) for passport in passports])
  print(part2)

File: 2020/4/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8", (err, data) => {
    if (err) {
      console.log(err);
    } else {
      return data;
    }
  })
  .split("\n");

const expectedFields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid", "cid"];
const yearPattern = /^\d{4}$/;
const hgtPattern = /^(\d+)(cm|in)$/;
const hclPattern = /^#[0-9a-f]{6}$/;
const eclPattern = /^(amb|blu|brn|gry|grn|hzl|oth)$/;
const pidPattern = /^\d{9}$/;

class Passport {
  constructor(fields) {
    this.fields = new Map();
    for (const [field, val] of fields) {
      this.fields.set(field, val);
    }
    this.validate1();
    this.validate2();
  }

  validate1() {
    const count = expectedFields.reduce((a, b) => a + this.fields.has(b), 0);
    this.isValid = count === 8 || (count === 7 && !this.fields.has("cid"));
  }

  validate2() {
    this.isValid2 = true;
    if (!this.isValid) {
      this.isValid2 = false;
      return;
    }
    if (!yearPattern.test(this.fields.get("byr"))) {
      this.isValid2 = false;
      return;
    }
    this.fields.set("byr", Number(this.fields.get("byr")));
    if (this.fields.get("byr") < 1920 || this.fields.get("byr") > 2002) {
      this.isValid2 = false;
      return;
    }
    if (!yearPattern.test(this.fields.get("iyr"))) {
      this.isValid2 = false;
      return;
    }
    this.fields.set("iyr", Number(this.fields.get("iyr")));
    if (this.fields.get("iyr") < 2010 || this.fields.get("iyr") > 2020) {
      this.isValid2 = false;
      return;
    }
    if (!yearPattern.test(this.fields.get("eyr"))) {
      this.isValid2 = false;
      return;
    }
    this.fields.set("eyr", Number(this.fields.get("eyr")));
    if (this.fields.get("eyr") < 2020 || this.fields.get("eyr") > 2030) {
      this.isValid2 = false;
      return;
    }
    const match = this.fields.get("hgt").match(hgtPattern);
    if (match === null) {
      this.isValid2 = false;
      return;
    }
    const hgt = Number(match[1]);
    if (match[2] === "cm" && (hgt < 150 || hgt > 193)) {
      this.isValid2 = false;
      return;
    }
    if (match[2] === "in" && (hgt < 59 || hgt > 76)) {
      this.isValid2 = false;
      return;
    }
    if (!hclPattern.test(this.fields.get("hcl"))) {
      this.isValid2 = false;
      return;
    }
    if (!eclPattern.test(this.fields.get("ecl"))) {
      this.isValid2 = false;
      return;
    }
    if (!pidPattern.test(this.fields.get("pid"))) {
      this.isValid2 = false;
      return;
    }
  }
}

if (input[input.length - 1].length > 0) {
  input.push("");
}

const fieldPattern = /(\S+):(\S+)/g;
let i = 0;
const passports = [];

while (i < input.length) {
  const currFields = [];
  while (input[i].length > 0) {
    const matches = input[i].matchAll(fieldPattern);
    for (const match of matches) {
      currFields.push([match[1], match[2]]);
    }
    i++;
  }
  passports.push(new Passport(currFields));
  i++;
}

const part1 = passports.reduce((a, b) => a + b.isValid, 0);
console.log(part1);
const part2 = passports.reduce((a, b) => a + b.isValid2, 0);
console.log(part2);
